lib/storage/http.js

```diff
@@ -15,20 +15,59 @@
 const http = require('@adobe/httptransfer');
 const { AssetComputeLogUtils, GenericError, RenditionTooLarge } = require('@adobe/asset-compute-commons');
 const { Action } = require('../action');
+const fs = require('fs-extra');
 
 const MAX_RETRY_DURATION_UPLOAD = 900000; // 15 mins
+const DEFAULT_MAX_CONCURRENT = 8;
+const DEFAULT_PREFERRED_PART_SIZE = 100 * 1024 * 1024; // Default part size is 100mb
+
+/**
+ * Get max concurrent value for node-httptransfer block transfer
+ *  - This is to make sure there is enough memory available inside the container for httptransfer to handle the block transfer
+ *  - If there is not enough memory available to handle the worst case (maxConcurrent * preferredPartSize), decrease maxConcurrent value
+ * @param {Number} preferredPartSize preferred part size for transferring chunks in MB
+ * @returns maxConcurrent integer value
+ */
+async function getMaxConcurrent(preferredPartSize) {
+    let containerMemorySize;
+    try {
+        // get container memory size
+        containerMemorySize = parseInt((await fs.readFile('/sys/fs/cgroup/memory/memory.limit_in_bytes')), 10) || undefined;
+    } catch (e) {
+        console.error(e);
+        // log error for reference but ignore
+    }
+    if (!Number.isFinite(containerMemorySize)) {
+        // if undefined or not a valid finite number, return default max concurrency
+        // this could happen if not running in the context of a docker container
+        return DEFAULT_MAX_CONCURRENT;
+    }
+    // total max memory that could be used at one time by node-httptransfer
+    const maxTotalMemory = preferredPartSize * DEFAULT_MAX_CONCURRENT;
+    const memoryNeeded = maxTotalMemory + (maxTotalMemory * 0.2); // add 20% memory buffer for other processing
+    if (containerMemorySize > memoryNeeded) {
+        return DEFAULT_MAX_CONCURRENT;
+    } else {
+        return Math.floor((containerMemorySize - (containerMemorySize * 0.2))/ preferredPartSize);
```

Comment:
You want to give 20% memory for other things, aka 80% for those down- up- loads, right? 
```suggestion
        return Math.floor((containerMemorySize * 0.8)/ preferredPartSize);
```

Reply:
yes, I kept it as is because I also factor in 20% buffer above when checking if there is enough memory: https://github.com/adobe/asset-compute-pipeline/blob/assets-211/lib/storage/http.js#L49

I did add more comments to make it more understandable